MapboxCoreNavigation/RouteProgress.swift

```diff
@@ -217,9 +223,43 @@ open class RouteProgress: NSObject {
             }
 
             congestionTravelTimesSegmentsByStep.append(congestionTravelTimesSegmentsByLeg)
+
+            maneuverCoordinateIndex = 0
+            if let segmentMaximumSpeedLimits = leg.segmentMaximumSpeedLimits {
+                for step in leg.steps {
+                    guard let coordinates = step.coordinates else { continue }
+                    let stepCoordinateCount = step.maneuverType == .arrive ? Int(step.coordinateCount) : coordinates.dropLast().count
+                    let nextManeuverCoordinateIndex = maneuverCoordinateIndex + stepCoordinateCount - 1
+
+                    guard nextManeuverCoordinateIndex < segmentMaximumSpeedLimits.count else { continue }
+
+                    let stepSegmentMaximumSpeedLimits = Array(segmentMaximumSpeedLimits[maneuverCoordinateIndex..<nextManeuverCoordinateIndex])
+                    maneuverCoordinateIndex = nextManeuverCoordinateIndex
+                    maximumSpeedLimitsByStep.append(stepSegmentMaximumSpeedLimits)
+                }
+            }
+
+            maximumSpeedLimitsByLeg.append(maximumSpeedLimitsByStep)
         }
     }
 
+    /**
+     Returns the SpeedLimit for the current position along the route. Returns SpeedLimit.invalid if the speed limit is unknown or missing.
+     */
+    public var currentSpeedLimit: SpeedLimit {
+        let coordinatesLeftOnStepCount = Int(floor((Double(currentLegProgress.currentStepProgress.step.coordinateCount)) * currentLegProgress.currentStepProgress.fractionTraveled))
```

Comment:
This is the same approach taken by #1242. Unfortunately, it assumes that each coordinate is evenly spaced along the step, which is almost never the case. For our purposes, we should assume that the segments are of arbitrary length.

With traffic congestion, it made sense to store congestion levels by segment, because we use those segments to create a shape layer all at once. Displaying the right speed limit at the right time is a very similar problem to presenting the right spoken or visual instruction at the right time, as RouteController and LegacyRouteController do. As you loop over the attributes, set each speed limit to the cumulative distance along the step. Then this computed property can become:

```swift
public var currentSpeedLimit: SpeedLimit {
    …
    let speedLimits = maximumSpeedLimitsByLeg[legIndex][currentLegProgress.stepIndex]
    return speedLimits.last { $0.distance <= distanceTraveled } ?? .invalid
}
```

Reply:
@1ec5 - Are you suggesting that I add a `distance` property to the SpeedLimit object or to keep a collection of (SpeedLimit, distance) pairs that is private to RouteProgress, built in the init() and referenced by currentSpeedLimit()?

Reply:
The latter, a collection of `(SpeedLimit, distance)` pairs. If the additional nesting turns out to be problematic, then the tuples could be grouped by leg and the distances could be measured from the start of the leg, but the tradeoff would be worse performance during the trip.